trusted/pal/src/pal.rs

```diff
@@ -0,0 +1,174 @@
+    /// deserializes and verifies the attestation (see `Host::verify_peer_attestation`),
+    /// uses the attestation to verify the data signatures, performs similar handshake
+    /// with the peer.
+    fn serialize(&self) -> Result<Vec<u8>>;
+
+    /// Signs a byte string using private signing key.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - The byte string to sign.
+    ///
+    /// # Returns
+    ///
+    /// An error if the attestation doesn't possess the signing capability (e.g. this
+    /// instance has been transferred over the wire and meant to be used for verification)
+    /// or if the signing failed, a success otherwise.
+    fn sign(&self, data: &[u8]) -> Result<Vec<u8>>;
+
+    /// Verifies signature of a byte string using public signing key.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - The byte string to verify signature for.
+    /// * `signature` - The signature of the byte string.
+    ///
+    /// # Returns
+    ///
+    /// An error if the signature cannot be verifed or is invalid, a success otherwise.
+    fn verify(&self, data: &[u8], signature: &[u8]) -> Result<()>;
+
+    /// Gets serialized public key used for signing by the trusted application.
+    /// The signing key is generated at the start of the trusted application and
+    /// remains unchanged through the lifetime.
+    ///
+    /// # Note
+    ///
+    /// Public signing key can be used to derive a trusted application identity.
+    /// For example, a trusted application that represents a node in Raft cluster
+    /// running inside of a group trusted hosts, it is important that Raft node
+    /// identity cannot be forged, does not require coordination to pick one and
+    /// has low collision chance. Hash of the public signing key is an identity
+    /// mechanism that is compliant with these requirements.
+    fn public_signing_key(&self) -> Vec<u8>;
+}
+
+/// Represents a trusted host, abstracting away the details of how trusted
+/// application is hosted (e.g. trusted host can be restricted kernel bare
+/// metal based encrypted virtual machine or a linux kernel docker based
+/// encrypted virtual machine).
+pub trait Host {
+    /// Gets attestation that represens a set of claims about the trusted application
+    /// and its host. The attestation is generated after the trusted application and
+    /// host are fully initialiezed, the attestation doesn't change throughout the
+    /// lifetime.
+    fn get_self_attestation(&self) -> Box<dyn Attestation>;
+
+    /// Gets serialized configuration that stays immutable through the lifetime of
+    /// the trusted application.
+    fn get_self_config(&self) -> Vec<u8>;
+
+    /// Sends messages through the communication channel that connects the trusted
+    /// application to the untrusted launcher.
+    ///
+    /// # Argumnets
+    ///
+    /// * `messages` - A set of messsages to send through the channel.
```

Comment:
Is it necessary to send multiple messages atomically? Would a simpler API that allows sending a single message not be sufficient? The application can always batch multiple actions in a single message if desired, so it seems unnecessary to add another level of grouping in this API.
Also I am hoping we could expose a raw [micro RPC transport](https://github.com/project-oak/oak/blob/8d2e369e8eda22eefd26eeef8994ff68b2541922/micro_rpc/src/lib.rs#L41-L49) to talk to the host, and the application will use that in whichever way it wants; would that be enough for your use case?

Reply:
Here send messages has a subtle semantic. Technically it doesn't send messages, it buffers them until the call to the receive_messages finishes. The application generates messages to be sent during the receive_messages call and then calls send_messages to be sent after receive_messages call completes. This is mainly due to the fact that in microRPC the call can only be issued from the Untrusted Launcher side. Hence, the send_messages buffers the messages to be sent and then pack them into a response for the microRPC.

With regards to the sending one or many messages at once, given the buffering semantic, it doesn't matter. However, processing by the application even of a single incoming message may result in generation of multiple messages (for example, a proposal message for the replicated log will result num replicas messages for appending entries).

Given that the purpose of this interface to abstract away the details of the platform (restricted kernel works with microRPC and Docker works with gRPC) the microRPC is the detail of how host (platform) is implemented.